pkg/tools/cli/loadtest/service/insert.go

```diff
@@ -17,28 +17,146 @@ package service
 
 import (
 	"context"
+	"sync"
 
+	"github.com/vdaas/vald/apis/grpc/agent/core"
 	"github.com/vdaas/vald/apis/grpc/gateway/vald"
 	"github.com/vdaas/vald/apis/grpc/payload"
+	"github.com/vdaas/vald/internal/errors"
 	"github.com/vdaas/vald/internal/net/grpc"
 	"github.com/vdaas/vald/pkg/tools/cli/loadtest/assets"
+	"github.com/vdaas/vald/pkg/tools/cli/loadtest/config"
 )
 
-func newInsert() (requestFunc, loaderFunc) {
-	return func(dataset assets.Dataset) ([]interface{}, error) {
-			vectors := dataset.Train()
-			ids := dataset.IDs()
-			requests := make([]interface{}, len(vectors))
-			for j, v := range vectors {
-				requests[j] = &payload.Object_Vector{
-					Id:     ids[j],
-					Vector: v,
-				}
+func insertRequestProvider(dataset assets.Dataset, batchSize int) (f func() interface{}, size int, err error) {
+	switch {
+	case batchSize == 1:
+		f, size = objectVectorProvider(dataset)
+	case batchSize >= 2:
+		f, size = objectVectorsProvider(dataset, batchSize)
+	default:
+		err = errors.New("batch size must be natural number.")
+	}
+	if err != nil {
+		return nil, 0, err
+	}
+	return f, size, nil
+}
+
+func objectVectorProvider(dataset assets.Dataset) (func() interface{}, int) {
```

Comment:
I am sorry if I made misstake.


Is the following simpler?
```go
func objectVectorProvider(dataset assets.Dataset) (func() interface{}, int) {
	v := dataset.Train()
	ids := dataset.IDs()
	var i int64 = -1
	return func() (ret interface{}) {
		if i := int(atomic.AddInt64(&i, 1)); i < len(v) {
			ret = &payload.Object_Vector{
				Id:     ids[i],
				Vector: v[i],
			}
		}
		return
	}, size
}

```

Reply:
Oh, that sounds good.
Is the **size** just missing isn't it?

Reply:
yes. The size does not change, so get it with len.